### PR TITLE
Add rule `jsx-classname-namespace` for enforcing CSS namespace guidelines

### PR DESCRIPTION
**Pull request by aduth**
This pull request seeks to introduce a new rule, `jsx-classname-prefix`, which produces an error when the `className` of a JSX element does not adhere to the namespace guidelines from the Calypso CSS documentation:

https://github.com/Automattic/wp-calypso/blob/master/docs/coding-guidelines/css.md

**Testing instructions:**
1. Clone repository and checkout branch `add/jsx-classname-prefix`
2. `npm install`
3. `npm ln`
4. Change to directory containing [wp-calypso repository](https://github.com/Automattic/wp-calypso.git)
5. `npm ln eslint-plugin-wpcalypso`
6. `./node_modules/.bin/eslint --rule 'wpcalypso/jsx-classname-namespace: 2' client/layout/index.jsx`
7. Note three errors are produced, corresponding with `wp-content`, `wp-primary`, and `wp-secondary` elements

``` text
/Users/andrew/Documents/Code/wp-calypso/client/layout/index.jsx
  200:23  error  Incorrect usage of CSS namespace  wpcalypso/jsx-classname-namespace
  205:24  error  Incorrect usage of CSS namespace  wpcalypso/jsx-classname-namespace
  206:26  error  Incorrect usage of CSS namespace  wpcalypso/jsx-classname-namespace
```

Ensure Mocha tests pass by running `npm test`.

/cc @mtias

### CODE REVIEW

**Comment by mtias** [comment]
Lovely :)

**Comment by lamosty** [comment]
Woot, pretty nice!

selection_024

**Comment by deBhal** [comment]
It would be nice if the error message pointed towards a solution, maybe something like 
`jsx className should follow calypso css coding guidelines` or
`className should follow wpcalypso.wordpress.com/devdocs/docs/coding-guidelines/css.md`

File: lib/rules/jsx-classname-namespace.js [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+					case 'ArrowFunctionExpression':
+						if ( 'VariableDeclarator' === functionExpression.parent.type ) {
+							functionName = functionExpression.parent.id.name;
+						}
+						break;
+				}
+
+				// If we've exhausted parent options, check to see whether
+				// program's exports refer to this expression
+				return functionName && 'Program' === parent.type && parent.body.some( function( programNode ) {
```
Comment (deBhal) [inline lib/rules/jsx-classname-namespace.js]: It's a bit pedantic, but when there's no functionName `isRootRenderedElement()` returns `undefined`, but it returns either `true` or `false` in every other case.  `functionName !== undefined` here would make for return values more consistent.

File: lib/rules/jsx-classname-namespace.js [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+			parent = parent.parent;
+		} while ( parent );
+
+		return false;
+	}
+
+	return {
+		JSXAttribute: function( node ) {
+			var literal, classNames, isError, isRoot, namespace;
```
Comment (deBhal) [inline lib/rules/jsx-classname-namespace.js]: "rawClassNames" might be a better name than "literal"

**Comment by deBhal** [comment]
I commented on a couple of cosmetic things, but LGTM 👍

**Comment by aduth** [comment]
Thanks all for taking a look, and for the code feedback @deBhal . I've addressed each of your concerns in the rebased 999148c. The implementation has been modified slightly, with a few more tests for edge cases I'd encountered in running this on Calypso's codebase reporting false positives.

I'd encourage taking this for a spin on the entire codebase and testing to see whether the reported errors are in line with expectations:

``` bash
cd /path/to/eslint-plugin-wpcalypso
git branch -D add/jsx-classname-prefix
git fetch
git checkout add/jsx-classname-prefix
npm install
npm ln
cd /path/to/wp-calypso
npm install
npm ln eslint-plugin-wpcalypso
./node_modules/.bin/eslint --quiet --ext .jsx --rule 'wpcalypso/jsx-classname-namespace: 2' client/
```

**Comment by deBhal** [comment]
I wanted to have a more thorough look at the output, so I added this patch:

``` diff
--- a/lib/rules/jsx-classname-namespace.js
+++ b/lib/rules/jsx-classname-namespace.js
@@ -165,13 +165,17 @@ var rule = module.exports = function( context ) {
            if ( isError ) {
                context.report( {
                    node: node,
-                   message: rule.ERROR_MESSAGE
+                   message: rule.ERROR_MESSAGE,
+                   data: {
+                       expected: isRoot ? namespace : namespace + '__*',
+                       actual: rawClassName.value
+                   }
                } );
            }
        }
    };
 };

-rule.ERROR_MESSAGE = 'className should follow CSS namespace guidelines';
+rule.ERROR_MESSAGE = 'className should follow CSS namespace guidelines.Expected to find "{{ expected }}" in "{{ actual }}"';

 rule.schema = [];
```

Now, that makes the error message way more useful, but I've not added it because it makes the error message _really_ long.  I think it's worth it, but it definitely needs a second opinion.

If you make the error message something like '\n\t\t' you get a slightly better effect, but what would be really awesome would be a formatter that could handle newlines.

For this PR, though, the outputs look good.  It seems a little over-enthusiastic for `client/layout/guided-tours/steps.js` where there are several closely related sub-components in the same file, but we can disable the rule there easily enough.

I'm sticking with the 👍  with or without the data change (don't forget to update the tests if you decide to go that way).

**Comment by aduth** [comment]
@deBhal : I like the idea of adding more context. The concern with adding longer messages (and especially multi-line messages) is that the messages are not always displayed in a terminal. For example, in my editor's integration, it shows in a very small status bar at the bottom of the editor:

image

As you might notice in that screenshot, I compromised and added the expected value, which I think will be helpful to many folks (see ae03027).

> For this PR, though, the outputs look good. It seems a little over-enthusiastic for `client/layout/guided-tours/steps.js`

Yeah, I anticipate this might lead into a bigger discussion on whether we should include multiple components in the same file like this. As you mentioned, it's not a common occurrence and shouldn't be a big concern to simply disable the rule for the file. Alternatively, we could explore expanding the logic here from a fairly basic "if is top-level element in `render`, assume it's root" to something more akin to the stateless function testing where we determine if the render only occurs as part of the default exported component.

The case I expect to be more controversial, and is intended for the moment, is elements that contain a general-purpose `className` (`screen-reader-text`, etc.) that don't also include its component-specific namespaced class.

**Comment by aduth** [comment]
Going to merge and publish this one. If need arises, we can iterate further in patch releases.

**Comment by aduth** [comment]
Aaaand I already found a false-positive:

``` js
<div>{ isOk && <div className="foo__child" /> }</div>
```

Incorrectly assumes that the child is the root element because its direct parent is not a JSX element. Will likely need to check all ancestors as being JSX element.

In other news, here's the pull request for Calypso: https://github.com/Automattic/wp-calypso/pull/6090
